File: solutions/python/19.py
def distribute_weight(weight: int) -> str:
    box_sizes = {
        10: [" _________ ", "|         |", "|_________|"],
        5: [" _____ ", "|     |", "|_____|"],
        2: [" ___ ", "|___|"],
        1: [" _ ", "|_|"]
    }
    box_counts = {}
    for size in box_sizes.keys():
        box_counts[size] = weight // size
        weight %= size
    boxes_list = []
    for size, quantity in box_counts.items():
        boxes_list = [size]*quantity + boxes_list
    result = ""
    result += "\n".join(box_sizes[boxes_list[0]])
    for i in range(len(boxes_list))[1:]:
        previous_box = boxes_list[i-1]
        box = boxes_list[i]
        if (previous_box != box):
            start_box_cover_index = len(box_sizes[previous_box][-1])
            result += box_sizes[box][0][start_box_cover_index:-1]
        result += "\n"
        result += "\n".join(box_sizes[box][1:])
    return result


if __name__ == "__main__":
    weights1 = distribute_weight(1)
    print(weights1)
    # Returns:
    #  _
    # |_|

    weights2 = distribute_weight(2)
    print(weights2)
    # Returns:
    #  ___
    # |___|

    weights3 = distribute_weight(3)
    print(weights3)
    # Returns:
    #  _
    # |_|_
    # |___|

    weights4 = distribute_weight(4)
    print(weights4)
    # Returns:
    #  ___
    # |___|
    # |___|

    weights5 = distribute_weight(5)
    print(weights5)
    # Returns:
    #  _____
    # |     |
    # |_____|

    weights6 = distribute_weight(6)
    print(weights6)
    # Returns:
    #  _
    # |_|___
    # |     |
    # |_____|

"""
  Score: # TODO: Code triggers error in web page
"""

File: solutions/ts/19.ts
type BoxSize = "1" | "2" | "5" | "10";

function distributeWeight(weight: number): string {
  const boxRepresentations: Record<BoxSize, string[]> = {
    1: [" _ ", "|_|"],
    2: [" ___ ", "|___|"],
    5: [" _____ ", "|     |", "|_____|"],
    10: [" _________ ", "|         |", "|_________|"],
  };
  const usedBoxes: Record<string, number> = { 1: 0, 2: 0, 5: 0, 10: 0 };
  let result = "";
  let remainder = weight;
  const piledBoxes: BoxSize[] = [];
  for (const size of Object.keys(usedBoxes).reverse()) {
    const quantity = Math.floor(remainder / parseInt(size));
    piledBoxes.unshift(...new Array(quantity).fill(size));
    usedBoxes[size] = quantity;
    remainder = remainder % parseInt(size);
  }
  result += boxRepresentations[piledBoxes[0]].join("\n");
  for (let i = 1; i < piledBoxes.length; i++) {
    const previousBox = piledBoxes[i - 1];
    const box = piledBoxes[i];
    if (previousBox !== box) {
      const startBoxCoverIndex = boxRepresentations[previousBox].at(-1)!.length;
      result += boxRepresentations[box][0].slice(startBoxCoverIndex, -1);
    }
    result += "\n";
    result += boxRepresentations[box].slice(1).join("\n");
  }
  return result;
}

const weights1 = distributeWeight(1);
console.log(weights1);
// Returns:
//  _
// |_|

const weights2 = distributeWeight(2);
console.log(weights2);
// Returns:
//  ___
// |___|

const weights3 = distributeWeight(3);
console.log(weights3);
// Returns:
//  _
// |_|_
// |___|

const weights4 = distributeWeight(4);
console.log(weights4);
// Returns:
//  ___
// |___|
// |___|

const weights5 = distributeWeight(5);
console.log(weights5);
// Returns:
//  _____
// |     |
// |_____|

const weights6 = distributeWeight(6);
console.log(weights6);
// Returns:
//  _
// |_|___
// |     |
// |_____|

/* 
  Score: ★★★★★
*/
